rpc/lib/client/http_client.go

```diff
@@ -28,67 +28,60 @@ const (
 	protoTCP   = "tcp"
 )
 
-// HTTPClient is a common interface for JSONRPCClient and URIClient.
-type HTTPClient interface {
-	Call(method string, params map[string]interface{}, result interface{}) (interface{}, error)
-	Codec() *amino.Codec
-	SetCodec(*amino.Codec)
+// Parsed URL structure
+type parsedURL struct {
+	url.URL
 }
 
-// protocol - client's protocol (for example, "http", "https", "wss", "ws", "tcp")
-// trimmedS - rest of the address (for example, "192.0.2.1:25", "[2001:db8::1]:80") with "/" replaced with "."
-func toClientAddrAndParse(remoteAddr string) (network string, trimmedS string, err error) {
-	protocol, address, err := parseRemoteAddr(remoteAddr)
+// Parse URL and set defaults (a flexible implementation for 'parseRemoteAddr')
+func newParsedURL(remoteAddr string) (*parsedURL, error) {
+	u, err := url.Parse(remoteAddr)
 	if err != nil {
-		return "", "", err
+		return nil, err
 	}
 
+	// default to tcp if nothing specified
+	if u.Scheme == "" {
+		u.Scheme = protoTCP
+	}
+
+	return &parsedURL{*u}, nil
+}
+
+// Change protocol to HTTP for unknown protocols and TCP protocol (flexible HTTPS downgrade fix) - useful for RPC connections
+func (u *parsedURL) SetDefaultSchemeHTTP() {
 	// protocol to use for http operations, to support both http and https
-	var clientProtocol string
-	// default to http for unknown protocols (ex. tcp)
-	switch protocol {
+	switch u.Scheme {
 	case protoHTTP, protoHTTPS, protoWS, protoWSS:
-		clientProtocol = protocol
+		// known protocols not changed
 	default:
-		clientProtocol = protoHTTP
+		// default to http for unknown protocols (ex. tcp)
+		u.Scheme = protoHTTP
 	}
+}
 
-	// replace / with . for http requests (kvstore domain)
-	trimmedAddress := strings.Replace(address, "/", ".", -1)
-	return clientProtocol, trimmedAddress, nil
+// Get full address without the protocol - useful for Dialer connections
+func (u parsedURL) GetHostWithPath() string {
+	// Remove protocol, userinfo and # fragment, assume opaque is empty
+	return u.Host + u.EscapedPath()
 }
 
-func toClientAddress(remoteAddr string) (string, error) {
-	clientProtocol, trimmedAddress, err := toClientAddrAndParse(remoteAddr)
-	if err != nil {
-		return "", err
-	}
-	return clientProtocol + "://" + trimmedAddress, nil
-}
-
-// network - name of the network (for example, "tcp", "unix")
-// s - rest of the address (for example, "192.0.2.1:25", "[2001:db8::1]:80")
-// TODO: Deprecate support for IP:PORT or /path/to/socket
-func parseRemoteAddr(remoteAddr string) (network string, s string, err error) {
-	parts := strings.SplitN(remoteAddr, "://", 2)
-	var protocol, address string
-	switch {
-	case len(parts) == 1:
-		// default to tcp if nothing specified
-		protocol, address = protoTCP, remoteAddr
-	case len(parts) == 2:
-		protocol, address = parts[0], parts[1]
-	default:
-		return "", "", fmt.Errorf("invalid addr: %s", remoteAddr)
-	}
+// Get a trimmed address (Implements previous 'toClientAddrAndParse') - useful for WS connections
```

Comment:
> (Implements previous 'toClientAddrAndParse')

do we really need to mention funcs that were removed

Reply:
Not really, this is again a backward-compatibility thing. I actually added these here so if someone wants to check backward compatibility, they have an easier time going through the code.

I'll remove it.